### PR TITLE
Make protobuf code generation not depend on system installation

### PR DESCRIPTION
**Pull request by nnmm**
Instead, protoc is expected to be contained within `target/protobuf/bin/protoc`. This eliminates confusion and brittleness around PATH ordering.

A follow-up PR will factor out the boilerplate of generating protobufs (setting the PATH to point to the above directory, wrapping in a `mod proto` etc.

### CODE REVIEW

**Comment by nnmm** [comment]
meow @tblaette 
fyi @feuerste

File: README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ It was build to serve the needs of the [Cartographer](https://github.com/googlec
 The project consist of a root crate that can build and read octrees on disk and viewer binaries to visualize the data. For Mac OS X, we assume below that you've installed [Homebrew](https://brew.sh).
 
 - Install Rust: `curl https://sh.rustup.rs -sSf | sh`. See <https://rustup.rs> for details.
-- Install protobuf > 3.0. See [ci/install_proto3.sh](https://github.com/googlecartographer/point_cloud_viewer/blob/master/ci/install_proto3.sh) for Linux or run `brew install protobuf` on Mac OS X.
+- Install `protoc` > 3.0 in `target/protobuf/bin`. The [`protobuf` Makefile target](https://github.com/googlecartographer/point_cloud_viewer/blob/master/Makefile) will compile `protoc` from source and place it there. Alternatively, you can symlink to a `protoc` installed on your system, e.g. from `brew install protobuf` on Mac OS X.
```
Comment (tblaette) [inline README.md]: make the separate steps for Linux vs. macOS clearer. Something along these lines:
- Linux: Install protoc by running the makefile
- macOS: brew install protobuf, symlink from ... to ...

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ROOT_DIR := $(abspath $(dir ${MAKEFILE_PATH}))
+PROTOBUF_DIR := ${ROOT_DIR}/target/protobuf
+
+.PHONY: protobuf
+protobuf: ${PROTOBUF_DIR}/protoc
+
+${PROTOBUF_DIR}/protoc:
+	mkdir -p ${PROTOBUF_DIR}
+	$(eval TMPDIR := $(shell mktemp -d))
+	git clone https://github.com/google/protobuf.git ${TMPDIR}/protobuf
```
Comment (tblaette) [inline Makefile]: check out a specific version of protobuf and not the latest

File: README.md (hunk 1 above)
Comment (nnmm) [reply README.md]: Do you agree with the current formulation? The script works for both Mac and Linux.

File: Makefile (hunk 2 above)
Comment (nnmm) [reply Makefile]: Done. Checking out 3.6.1.

**Comment by nnmm** [comment]
@wally-the-cartographer merge

File: point_viewer_grpc_proto_rust/build.rs [hunk 3]
```diff
@@ -48,12 +48,15 @@ fn main() {
     println!("cargo:rerun-if-changed=src/proto.proto");
 
     let git_repo_root = find_git_repo_root();
+    let old_path = env::var("PATH");
```
Comment (feuerste) [inline point_viewer_grpc_proto_rust/build.rs]: Would it make sense to have a small RAII helper for temporarily changing the path and use it everywhere, something like?
```rust
struct TempPath {
    old_path: String,
}

impl TempPath {
    pub fn new(new_path: &Str) -> Self {
        let old_path = env::var("PATH");
        env::set_var("PATH", new_path);
        TempPath{ old_path }
    }
}

impl Drop for TempPath {
    fn drop(&mut self) {
        env::set_var("PATH", self.old_path.unwrap_or("".to_string()));
    }
}
```

Comment (nnmm) [reply point_viewer_grpc_proto_rust/build.rs]: Something like that, yes. I'd do that in a separate `build_proto` crate that we can use in build scripts for other crates.

Comment (feuerste) [reply point_viewer_grpc_proto_rust/build.rs]: Cool, thanks

**Comment by nnmm** [comment]
@cschuet Could you merge this PR?

**Comment by nnmm** [comment]
Alternatively @gaschler?

**Comment by nnmm** [comment]
@sebastianklose Could you merge and/or point me to the right person to ping about Wally?

**Comment by nnmm** [comment]
Thank you, Andre!

**Pull request merged**